src/api/pool.go

```diff
@@ -0,0 +1,109 @@
+package api
+
+import (
+	"context"
+	"github.com/go-resty/resty/v2"
+	"golang.org/x/sync/errgroup"
+	"runtime"
+	"sync"
+)
+
+type Pool struct {
+	client          *Client         // http client
+	ctx             context.Context // context of the parallel work
+	workers         *errgroup.Group // error group -> if one worker fails, all will be stopped
+	counter         sync.WaitGroup  // detect when all requests are processed
+	sendersCount    int
+	processorsCount int
+	processor       func(pool *Pool, response *resty.Response) error
+	requests        chan *resty.Request
+	responses       chan *resty.Response
+}
+
+func (c *Client) NewPool(processor func(pool *Pool, response *resty.Response) error) *Pool {
+	workers, ctx := errgroup.WithContext(c.parentCtx)
+	return &Pool{
+		client:          c,
+		ctx:             ctx,
+		workers:         workers,
+		counter:         sync.WaitGroup{},
+		sendersCount:    MaxIdleConns,
+		processorsCount: runtime.NumCPU(),
+		processor:       processor,
+		requests:        make(chan *resty.Request, 500),
+		responses:       make(chan *resty.Response),
+	}
+}
+
+// R creates request
+func (p *Pool) R(method string, url string) *resty.Request {
+	r := p.client.R().SetContext(p.ctx)
+	r.Method = method
+	r.URL = url
+	return r
+}
+
+// Add request to pool
+func (p *Pool) Add(request *resty.Request) {
+	p.counter.Add(1)
+	p.requests <- request
+}
+
+// Wait until all requests done
+func (p *Pool) Wait() error {
+	return p.workers.Wait()
+}
+
+func (p *Pool) Start() {
+	// Close channels when all requests are processed
+	p.workers.Go(func() error {
+		defer close(p.requests)
+		defer close(p.responses)
+		p.counter.Wait()
```

Comment:
Toto vlakno caka, kym bude na pocitadle `0`, `p.counter.Wait()` a potom uzatvori kanal `requests` a `response` .... ked sa uzatvori kanal ... tak sa ukoncia `for` loops ktore su nizsie .... a preto sa ukonci aj vykonavanie vsetkych `goroutines` a odblokuje sa `p.workers.Wait()`, ktory je v kode vyssie.